### PR TITLE
Filter Grammar

### PR DESCRIPTION
**Pull request by YuriHechter**
https://docs.google.com/spreadsheets/d/1zHVo8NgQU_qjkRyrDxn2Q5bWFjjIEEm92YTuDZtocvs/edit#gid=1454852817

### CODE REVIEW

File: cache/search_stats.go [hunk 1]
```diff
@@ -141,6 +141,8 @@ type SearchStatsCache interface {
 	// |holiday| is the UID of the tag that is children of 'holidays' tag
 	DoesHolidayExist(holiday string, year string) bool
 	DoesHolidaySingle(holiday string, year string) bool
+
+	DoesSourceTitleWithMoreThanOneWordExist(title string) bool
```
Comment (YuriHechter) [inline cache/search_stats.go:145]: add comment with problematic source names example

File: search/engine.go [hunk 2]
```diff
@@ -651,24 +675,13 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
...
-		resultTypes = consts.ES_SEARCH_RESULT_TYPES
-	}
-
 	intents, err := e.AddIntents(&query, preference, consts.INTENTS_SEARCH_COUNT, sortBy)
 	if err != nil {
 		log.Errorf("ESEngine.DoSearch - Error adding intents: %+v", err)
 	}
 
 	// Filter out duplicates of regular results and intents carousel results
+	// Note: Currently we don't have such logic for 'filter grammar' results.
```
Comment (bbfsdev) [inline search/engine.go:684]: Don't understand this comment.

File: search/engine.go [hunk 3]
```diff
@@ -762,6 +788,49 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 		resultsByLang[lang] = append(resultsByLang[lang], tweets)
 	}
 
+	filteredByLang := <-grammarsFilteredResultsByLangChannel
```
Comment (bbfsdev) [inline search/engine.go:791]: Comment what this loop does... Consider extracting those loops to functions with documentation (same as the comment before the loop).

File: search/grammar_v2_engine.go [hunk 4]
```diff
@@ -399,3 +487,61 @@ func (e *ESEngine) collectionHitFromSql(query string) (*elastic.SearchHit, error
 	}
 	return hit, nil
 }
+
+func retrieveTextVarValues(str string) []string {
```
Comment (bbfsdev) [inline search/grammar_v2_engine.go:491]: Document this function, what it does... provide example and/or general rule.

File: search/grammar_v2_engine.go [hunk 5]
```diff
@@ -399,3 +487,61 @@ func (e *ESEngine) collectionHitFromSql(query string) (*elastic.SearchHit, error
...
+			}
+			filtered = make([]rune, 0)
+		} else if !inHighlight {
+			filtered = append(filtered, r)
+		}
+	}
+	return textVarValues
+}
+
+func (e *ESEngine) filterSearch(requests []*elastic.SearchRequest) ([]*elastic.SearchResult, map[string]bool, *float64, error) {
```
Comment (bbfsdev) [inline search/grammar_v2_engine.go:516]: What that do?

File: cache/search_stats.go (hunk 1 above)
Comment (YuriHechter) [reply cache/search_stats.go:145]: comment added to SearchGrammarsV2 function

File: search/engine.go (hunk 2 above)
Comment (YuriHechter) [reply search/engine.go:684]: When we have a lessons carousel we filter out the regular results that are also exist in the carousel . Currently we don't support showing intent carousels for grammar filtered results. So we are also not filtering many appearances of lesson results.

File: search/engine.go (hunk 3 above)
Comment (YuriHechter) [reply search/engine.go:791]: comment added

File: search/grammar_v2_engine.go (hunk 4 above)
Comment (YuriHechter) [reply search/grammar_v2_engine.go:491]: comments added

File: search/grammar_v2_engine.go (hunk 5 above)
Comment (YuriHechter) [reply search/grammar_v2_engine.go:516]: Results search according to content type filter. This logic is separated from the SearchGrammarsV2 function to minimize the size of SearchGrammarsV2  function .

File: search/engine.go (hunk 2 above)
Comment (bbfsdev) [reply search/engine.go:684]: Copy paste your explanation into the comment

File: search/grammar_v2_engine.go (hunk 5 above)
Comment (bbfsdev) [reply search/grammar_v2_engine.go:525]: Add comment.

File: search/engine.go (hunk 2 above)
Comment (YuriHechter) [reply search/engine.go:684]: added

File: search/grammar_v2_engine.go (hunk 5 above)
Comment (YuriHechter) [reply search/grammar_v2_engine.go:525]: Added

**Review by bbfsdev — APPROVED** [review]

**Pull request merged**